### PR TITLE
[FLINK-14912][Table] create, drop catalog functions through catalog manager

### PR DESCRIPTION
**Pull request by HuangZhenQiu**
## What is the purpose of the change
Create, drop and alter catalog functions through catalog manager.

## Brief change log

  - Add create, drop, alter function operation
  - Add SqlNode to Operation conversion logic SqlToOperationConverter
  - Add the catalog function registration and drop logic in table envrionments
  - add CatalogFunctionITCase for end to end testing function DDL to create and drop catalog function.

## Verifying this change

It is end to end tested in new added CatalogFunctionITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by HuangZhenQiu** [comment]
@bowenli86 @xuefuz Please review the PR at your most convenient time. Currently, only SqlCreateFunction and SqDropFunction are added in SqlKind, but none of them is actually included in the DDL enum. In this case, I have to change the calcite first, then wait for the new calcite release. Do you have a better strategy to handle it?

**Comment by bowenli86** [comment]
> @bowenli86 @xuefuz Please review the PR at your most convenient time. Currently, only SqlCreateFunction and SqDropFunction are added in SqlKind, but none of them is actually included in the DDL enum. In this case, I have to change the calcite first, then wait for the new calcite release. Do you have a better strategy to handle it?

I'm not familiar with calcite. @danny0405 @KurtYoung what do you think?

**Comment by danny0405** [comment]
Thanks for your nice job @HuangZhenQiu ~ Reviewing now ~

**Comment by danny0405** [comment]
Do you mean CALCITE-3349 ? I think we do not need to wait for CALCITE-3349 in the release, we can use `Sqlkind.OTHER_DDL` instead first. After all, we should always implement the `SqlCreateFunction` and `SqlDropFunction` in flink-sql-parser. We can use the DDL as `SqlKind` if we upgrade to Calcite 1.22.0.(You should logger a JIRA issue under the 1.22 version Calcite upgrade umbrella issue).

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 1]
```diff
@@ -474,13 +478,23 @@ public void sqlUpdate(String stmt) {
...
+		} else if (operation instanceof AlterFunctionOperation) {
+			AlterFunctionOperation alterFunctionOperation = (AlterFunctionOperation) operation;
+			alterCatalogFunction(alterFunctionOperation);
+		} else if (operation instanceof DropFunctionOperation) {
+			DropFunctionOperation dropFunctionOperation = (DropFunctionOperation) operation;
+			dropCatalogFunction(dropFunctionOperation);
 		} else if (operation instanceof UseCatalogOperation) {
-			UseCatalogOperation useCatalogOperation = (UseCatalogOperation) operation;
-			catalogManager.setCurrentCatalog(useCatalogOperation.getCatalogName());
+				UseCatalogOperation useCatalogOperation = (UseCatalogOperation) operation;
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: revert?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/CatalogFunctionUtils.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	 * Converts language string the FunctionLanguage.
+	 * @param languageString  the language string from SQL parser
+	 * @return supported FunctionLanguage otherwise raise UnsupportedOperationException.
+	 */
+	public static FunctionLanguage parseLanguage(String languageString) {
+		if (StringUtils.isNullOrWhitespaceOnly(languageString)) {
+			return FunctionLanguage.JAVA;
+		}
+
+		FunctionLanguage language = FunctionLanguage.valueOf(languageString);
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/CatalogFunctionUtils.java]: IIRC, if the value is unmatched, an exception will be thrown, and language won't be null.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/CatalogFunctionUtils.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	 * @param functionIdentifier function identifier
+	 * @param required whether catalog has to be exist
+	 * @return return catalog otherwise throw UnsupportedOperationException.
+	 */
+	public static Optional<Catalog> lookupCatalog(
+		CatalogManager catalogManager,
+		ObjectIdentifier functionIdentifier,
+		boolean required) {
+		Optional<Catalog> catalogOptional = catalogManager.getCatalog(functionIdentifier.getCatalogName());
+		if (!catalogOptional.isPresent() && required) {
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/CatalogFunctionUtils.java]: `ignoreifnotexist` is for function, not catalog/db, right?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 4]
```diff
@@ -622,6 +636,50 @@ private void registerTableSinkInternal(String name, TableSink<?> tableSink) {
 			.map(CatalogManager.TableLookupResult::getTable);
 	}
 
+	private void createCatalogFunction(CreateFunctionOperation createFunctionOperation) {
+		Optional<Catalog> catalogOptional = CatalogFunctionUtils.lookupCatalog(
+			catalogManager, createFunctionOperation.getFunctionIdentifier(), !createFunctionOperation.isIgnoreIfExists());
+		if (catalogOptional.isPresent()) {
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: `ignoreifnotexist` is for function, not catalog/db, right?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/CatalogFunctionUtils.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import org.apache.flink.table.catalog.ObjectIdentifier;
+import org.apache.flink.util.StringUtils;
+
+import java.util.Optional;
+
+/**
+ * Utilities for catalog function.
+ */
+@Internal
+public final class CatalogFunctionUtils {
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/CatalogFunctionUtils.java]: feels this class is not necessary to have

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 6]
```diff
@@ -159,6 +175,47 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
 		return new DropTableOperation(identifier, sqlDropTable.getIfExists());
 	}
 
+	/** Convert CREATE FUNCTION statement. */
+	private Operation convertCreateFunction(SqlCreateFunction sqlCreateFunction) {
+		FunctionLanguage language = CatalogFunctionUtils.parseLanguage(sqlCreateFunction.getFunctionLanguage());
```
Comment (bowenli86) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java]: just put `parseLanguange` in SqlToOperationConverter.java ? I feel flink and blink planner can decide whether they support python function independently

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java [hunk 7]
```diff
@@ -166,6 +182,48 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
 		return new DropTableOperation(identifier, sqlDropTable.getIfExists());
 	}
 
+
+	/** Convert CREATE FUNCTION statement. */
+	private Operation convertCreateFunction(SqlCreateFunction sqlCreateFunction) {
+		FunctionLanguage language = CatalogFunctionUtils.parseLanguage(sqlCreateFunction.getFunctionLanguage());
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java]: just put parseLanguange in SqlToOperationConverter.java ? I feel flink and blink planner can decide whether they support python function independently

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala [hunk 8]
```diff
@@ -613,24 +617,74 @@ abstract class TableEnvImpl(
       .map(_.getTable)
   }
 
+  private def createCatalogFunction(createFunctionOperation: CreateFunctionOperation)= {
+    val objectIdentifier = createFunctionOperation.getFunctionIdentifier
+    val catalogOptional = CatalogFunctionUtils.lookupCatalog(
+      catalogManager, objectIdentifier, !createFunctionOperation.isIgnoreIfExists)
+
+    if (catalogOptional.isPresent) {
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala]: ditto. `ignoreIfExists` is for function, not catalog/db

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala [hunk 9]
```diff
@@ -613,24 +617,74 @@ abstract class TableEnvImpl(
...
+      }
+    }
+  }
+
+  private def alterCatalogFunction(alterFunctionOperation: AlterFunctionOperation) = {
+    val objectIdentifier = alterFunctionOperation.getFunctionIdentifier
+    val catalogOptional = CatalogFunctionUtils.lookupCatalog(
+      catalogManager, objectIdentifier, !alterFunctionOperation.isIfExists)
+
+    if (catalogOptional.isPresent) {
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala]: ditto

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala [hunk 10]
```diff
@@ -613,24 +617,74 @@ abstract class TableEnvImpl(
...
+      }
+    }
+  }
+
+  private def dropCatalogFunction(dropFunctionOperation: DropFunctionOperation) = {
+    val objectIdentifier = dropFunctionOperation.getFunctionIdentifier
+    val catalogOptional = CatalogFunctionUtils.lookupCatalog(
+      catalogManager, objectIdentifier, !dropFunctionOperation.isIfExists)
+
+    if (catalogOptional != null) {
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala]: ditto

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala [hunk 11]
```diff
@@ -117,7 +117,9 @@ class FlinkPlannerImpl(
       }
       // no need to validate row type for DDL and insert nodes.
       if (sqlNode.getKind.belongsTo(SqlKind.DDL)
-        || sqlNode.getKind == SqlKind.INSERT) {
+        || sqlNode.getKind == SqlKind.INSERT
+        || sqlNode.getKind == SqlKind.CREATE_FUNCTION
+        || sqlNode.getKind == SqlKind.DROP_FUNCTION) {
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/FlinkPlannerImpl.scala]: miss ALTER_FUNCTION?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/catalog/CatalogFunctionITCase.scala [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+import org.junit.{Before, Test}
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+import org.hamcrest.CoreMatchers.containsString
+import org.hamcrest.MatcherAssert.assertThat
+
+/** Test cases for catalog function. */
+@RunWith(classOf[Parameterized])
+class CatalogFunctionITCase(isStreaming: Boolean) {
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/catalog/CatalogFunctionITCase.scala]: we need some IT case to actually create function and then invoke it in a query and verify output

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 1 above)
Comment (HuangZhenQiu) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: Typo. Thanks.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateFunctionOperation.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+ */
+public class CreateFunctionOperation implements CreateOperation {
+	private final ObjectIdentifier functionIdentifier;
+	private CatalogFunction catalogFunction;
+	private boolean ignoreIfExists;
+
+	public CreateFunctionOperation(
+		ObjectIdentifier functionIdentifier,
+		CatalogFunction catalogFunction,
+		boolean ignoreIfExists) {
```
Comment (danny0405) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateFunctionOperation.java]: Fix the indentation.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 14]
```diff
@@ -179,6 +195,72 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+
+		return new DropFunctionOperation(
+			identifier,
+			sqlDropFunction.getIfExists());
+	}
+
+	/**
+	 * Converts language string the FunctionLanguage.
+	 * @param languageString  the language string from SQL parser
+	 * @return supported FunctionLanguage otherwise raise UnsupportedOperationException.
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java]: `Converts language string` -> `Converts language string to`, also keep an empty line between the doc and `@param`.

You can use `@throws` to mark the probable exception.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 15]
```diff
@@ -179,6 +195,72 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
 		return new DropTableOperation(identifier, sqlDropTable.getIfExists());
 	}
 
+	/** Convert CREATE FUNCTION statement. */
+	private Operation convertCreateFunction(SqlCreateFunction sqlCreateFunction) {
+		FunctionLanguage language = parseLanguage(sqlCreateFunction.getFunctionLanguage());
+		CatalogFunction catalogFunction = new CatalogFunctionImpl(
+			sqlCreateFunction.getFunctionClassName().toValue(), language);
+
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java]: We better use `SqlLiteral.getValueAs` instead of `SqlLiteral.toValue`.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 16]
```diff
@@ -179,6 +195,72 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+			sqlDropFunction.getIfExists());
+	}
+
+	/**
+	 * Converts language string the FunctionLanguage.
+	 * @param languageString  the language string from SQL parser
+	 * @return supported FunctionLanguage otherwise raise UnsupportedOperationException.
+	 */
+	private FunctionLanguage parseLanguage(String languageString) {
+		if (StringUtils.isNullOrWhitespaceOnly(languageString)) {
```
Comment (danny0405) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java]: Move all the tool methods to the `//~ Tools` comment part.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala [hunk 17]
```diff
@@ -669,24 +673,67 @@ abstract class TableEnvImpl(
...
     functionCatalog
   }
 
-  private[flink] def getParserConfig: SqlParser.Config = planningConfigurationBuilder
+  private[flink] def getParserConfig = planningConfigurationBuilder
     .getSqlParserConfig
 
   /** Returns the Calcite [[FrameworkConfig]] of this TableEnvironment. */
   @VisibleForTesting
-  private[flink] def getFlinkPlanner: FlinkPlannerImpl = {
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/internal/TableEnvImpl.scala]: Unnecessary change ?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+	@Test
+	public void testCreateFunctionWithFullPath() {
+		String ddl1 = "create function default_catalog.default_database.f2 as" +
+			" 'org.apache.flink.function.TestFunction'";
+		tableEnv.sqlUpdate(ddl1);
+		assertTrue(Arrays.asList(tableEnv.listFunctions()).contains("f2"));
+
+		tableEnv.sqlUpdate("DROP FUNCTION IF EXISTS default_catalog.default_database.f2");
+		assertTrue(!Arrays.asList(tableEnv.listFunctions()).contains("f2"));
+	}
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java]: Use `assertFalse` instead. BTW, the test name is "testCreate" but you also test "Drop Function".

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+		tableEnv.sqlUpdate(ddl1);
+		assertTrue(Arrays.asList(tableEnv.listFunctions()).contains("f1"));
+
+		tableEnv.sqlUpdate("DROP FUNCTION IF EXISTS default_catalog.default_database.f1");
+		assertTrue(!Arrays.asList(tableEnv.listFunctions()).contains("f1"));
+	}
+
+	@Test
+	public void testCreateFunctionWithFullPath() {
+		String ddl1 = "create function default_catalog.default_database.f2 as" +
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java]: We also need to test create function with `db_path.function_name`, each positive test should have a negative counterpart.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+			row.setField(i, objects[i]);
+		}
+
+		return row;
+	}
+
+	/**
+	 * Test udf class.
+	 */
+	public static class TestUDF extends ScalarFunction {
```
Comment (danny0405) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java]: We may also need to add test cases in blink-planner.

**Comment by danny0405** [comment]
Thank for the PR @HuangZhenQiu , the PR overall looks good, can you also add the cases for Blink planner ? Even though the code path looks the same, we still need the tests to make sure the logic works correctly.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateFunctionOperation.java (hunk 13 above)
Comment (HuangZhenQiu) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/CreateFunctionOperation.java]: Done

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java (hunk 16 above)
Comment (HuangZhenQiu) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java]: I assume you are suggesting to move all of the private function under //~Tools.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java (hunk 19 above)
Comment (HuangZhenQiu) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java]: Sure. Added accordingly.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java (hunk 18 above)
Comment (HuangZhenQiu) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java]: Done.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java (hunk 20 above)
Comment (HuangZhenQiu) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/catalog/CatalogFunctionTestBase.java]: Agree. Added.

**Comment by HuangZhenQiu** [comment]
@bowenli86 @danny0405 @xuefuz I resolved your comments. In the meantime, CatalogFunctionITCase is also added into blink planner. Please take one more round of review.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java [hunk 21]
```diff
@@ -239,7 +299,7 @@ private Operation convertCreateDatabase(SqlCreateDatabase sqlCreateDatabase) {
 		String databaseName = (fullDatabaseName.length == 1) ? fullDatabaseName[0] : fullDatabaseName[1];
 		boolean ignoreIfExists = sqlCreateDatabase.isIfNotExists();
 		String databaseComment = sqlCreateDatabase.getComment()
-												.map(comment -> comment.getNlsString().getValue()).orElse(null);
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java]: revert?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java [hunk 22]
```diff
@@ -258,10 +318,10 @@ private Operation convertDropDatabase(SqlDropDatabase sqlDropDatabase) {
 		String catalogName = (fullDatabaseName.length == 1) ? catalogManager.getCurrentCatalog() : fullDatabaseName[0];
 		String databaseName = (fullDatabaseName.length == 1) ? fullDatabaseName[0] : fullDatabaseName[1];
 		return new DropDatabaseOperation(
-				catalogName,
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java]: revert?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java [hunk 23]
```diff
@@ -287,14 +347,12 @@ private Operation convertAlterDatabase(SqlAlterDatabase sqlAlterDatabase) {
 		}
 		// set with properties
 		sqlAlterDatabase.getPropertyList().getList().forEach(p ->
-							properties.put(((SqlTableOption) p).getKeyString().toLowerCase(),
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/sqlexec/SqlToOperationConverter.java]: revert?

**Comment by HuangZhenQiu** [comment]
@bowenli86 I changed my local code for each of those lines. I didn't touch them.
It is in the right format from beginning rather than I did any change.

**Comment by HuangZhenQiu** [comment]
@danny0405 @bowenli86 The PR attached to the wrong ticket due to I mistakenly used wrong ID. I created another PR to correct it.
